### ISSUE TITLE
Roadmap to support e10s/Electrolysis and Firefox 57

### ISSUE THREAD

**Issue by brudgers** [kazu-yamamoto/Firemacs#36]
Kazu-yamamoto, thank you for making such a wonderful tool. I enjoy using Firemacs every day. I am posting this as an issue because I could not find another contact point.

Mozilla has decided to change the plugin architecture to facilitate sandboxing and other features. 

https://wiki.mozilla.org/Electrolysis

At the end of this year, Firefox 57 will be released without support for extensions using other technologies and without action, Firemacs will no longer be an option. It would be wonderful to see the program updated. 

I can understand if support for Firemacs no longer engages your attention. I have created the Github organization https://github.com/firemacs/Firemacs as a possible alternative should that be the case. 

If your plan is to transition Firemacs to e10s, I would be happy to see the Firemacs organization disappear and provide whatever help I can. All I really want is the ability to use Firemacs over the foreseeable future.

Thank you again, for making something I enjoy so much

**Comment by kazu-yamamoto** [comment]
@brudgers First of all, thank you for registering this issue.
Unfortunately, I don't have time to maintain Firemacs recently.
So, I would be happy if other people took over Firemacs.
Which do you think better, using the organization or giving the collaborator right to you?

**Comment by aaron-em** [comment]
Hello! I'm Aaron. Ben (@brudgers)  and I have chatted briefly on this subject, and I intend at this time to assume the development role for an Electrolysis-compatible version of Firemacs. I haven't previously had the pleasure of using the present version of the add-on, but having discovered it thanks to Ben's mentioning on Hacker News that he uses it, I find myself delighted with the concept and anxious to enjoy it in versions of Firefox which no longer gracefully support XUL-based add-ons. Thank you so much for your work on Firemacs to date!

Given the options at hand, I think that if all others involved concur, I would prefer development proceed under the new Firemacs organization. It does not seem likely beneficial to the project that we proceed in a repository whose owner may be hampered in maintenance actions by time zone differences and prior commitments. Nor would we wish to presume further than necessary upon your already constrained time.

Based on our prior conversation, I imagine that @brudgers would intend to proceed in a role more concerned with administering the organization and maintaining repositories, as indeed evidenced by his impressive initiative thus far. It seems therefore a natural enough fit that I would take on the role of actually migrating Firemacs into Firefox's new extension API, a task whose potential scope initially inspired temerity but which, as I've considered it further, I've found more and more appealing to tackle.

**Comment by brudgers** [comment]
@kazu-yamamoto I want to avoid doing any injustice toward you and your
creative work. I wish for you to be happy with the way Firemacs moves
forward. I completely understand not having time and I appreciate how your
work has made me happier. Your opinion matters to me.

One of the features of a Github organization is flexibility for
transferring project administration.   On the other hand, there may be
potential hurdles with Mozilla's administration regarding placement as a
Firefox Add-on if the project moves forward as an organization.  But I
don't know because I am in unfamiliar territory.

Over the long run, I think flexibility regarding administrative succession
improves the odds of project sustainability. An organization means the
project can be independent of my participation, your participation, etc. so
long as there is a community.

**Comment by brudgers** [comment]
@kazu-yamamoto Just to be clear, the way I envision the organization is
such that you may hold whatever roles and responsibilities you wish.

**Comment by kazu-yamamoto** [comment]
OK. The organization approach is appealing. So, just go ahead. :-)

**Comment by Petteri** [comment]
Firefox Quantum is released today. I can see that firemacs is not compatible: https://addons.mozilla.org/fi/firefox/addon/firemacs/

Any progress on this issue?

**Comment by aaron-em** [comment]
Unfortunately not. The blockers we identified [back in June](https://github.com/firemacs/Firemacs/issues/1#issuecomment-308237683) didn't get prioritized highly enough for fixes to land in 57, and while it would be possible to provide incomplete functionality in advance of those fixes, I'm really not interested in shipping code that's born broken and then being on the hook to deal with the entirely understandable yet completely unresolvable frustration of everyone thereafter who tries to use it and finds it's broken.

I'm following both the bugs mentioned there, and as soon as they make sufficient progress for a functional Firemacs web extension to be implemented, I'll begin doing that thing. I'd love to be able to offer an ETA on that - most of the reason why I involved myself in this in the first place was because I want to use Firemacs, too! But the closest I can get right now is "not long after the blockers go away". When that happens is up to Mozilla, and I get the sense that the people who might pick up these issues have many bigger fish to fry at this time.

**Comment by Petteri** [comment]
Thanks for the update and explanation. Hopefully Mozilla will fix those bugs soon!

**Comment by brudgers** [comment]
It is more a matter of nobody in the organization working on updating the
software. Myself included.

As far as my research went, there are only a couple of calls to the
obsolete interface. The rest is reorganizing the code to the new idiom (see
the design notes).

In the end, Firemacs is more or less:

1. A Javascript event listener that listens to the keyboard.

2. An configuration interface.



Ben

**Comment by aaron-em** [comment]
Yes. It's the event listener that is the problem - specifically, there being no way to apply it to the browser chrome as a whole, instead of having to inject it as a content script into tabs, and having it not work at all in those tabs whose content doesn't support such injection. I don't feel it is worth sinking more than PoC-level effort into this project until that gets resolved. Because if it never _does_ get resolved, then any effort invested is wasted - absent that fix, there's nothing I can do to produce a result that I'll be satisfied with, to say nothing of the frustration it'd be guaranteed to engender among users who find its shortcomings as unsatisfactory as I.

Is there something I'm missing here? I've been through the WebEx documentation pretty comprehensively, but that was a few months ago, and the proof-of-concept work I did is likewise a little way now in the past. Or maybe I'm following the wrong bugs or something - the point is, while as far as I can tell there is no way to make a Firemacs web extension work acceptably right now, I'm totally prepared to concede that there's a way after all and I just need help finding it. So if there's an API you can point me to that will allow me to bind event listeners on browser windows instead of in tab content, or otherwise get past the problem, then please point me to it!
